<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Simulating-linked-markers • CKMRsim</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Simulating-linked-markers">
<meta property="og:description" content="CKMRsim">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CKMRsim</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.2.999</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/CKMRsim-developer-notes.html">CKMRsim-developer-notes</a>
    </li>
    <li>
      <a href="../articles/CKMRsim-example-1.html">CKMRsim-example-1</a>
    </li>
    <li>
      <a href="../articles/CKMRsim-example-2-microsatellites.html">CKMRsim-example-2-microsatellites</a>
    </li>
    <li>
      <a href="../articles/CKMRsim-pairwise-relationships.html">CKMRsim-pairwise-relationships</a>
    </li>
    <li>
      <a href="../articles/CKMRsim-simulating-linked-markers.html">Simulating-linked-markers</a>
    </li>
    <li>
      <a href="../articles/CKMRsim-writing-geno-error-funcs.html">CKMRsim-writing-geno-error-functions</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Simulating-linked-markers</h1>
            
      
      
      <div class="hidden name"><code>CKMRsim-simulating-linked-markers.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">CKMRsim</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span>
<span class="co">#&gt; ── <span style="font-weight: bold;">Attaching packages</span> ─────────────────────────────────────── tidyverse 1.3.1 ──</span>
<span class="co">#&gt; <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">ggplot2</span> 3.3.6     <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">purrr  </span> 0.3.4</span>
<span class="co">#&gt; <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">tibble </span> 3.1.7     <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">dplyr  </span> 1.0.9</span>
<span class="co">#&gt; <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">tidyr  </span> 1.2.0     <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">stringr</span> 1.4.0</span>
<span class="co">#&gt; <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">readr  </span> 2.1.2     <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">forcats</span> 0.5.1</span>
<span class="co">#&gt; ── <span style="font-weight: bold;">Conflicts</span> ────────────────────────────────────────── tidyverse_conflicts() ──</span>
<span class="co">#&gt; <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">dplyr</span>::<span style="color: #00BB00;">filter()</span> masks <span style="color: #0000BB;">stats</span>::filter()</span>
<span class="co">#&gt; <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">dplyr</span>::<span style="color: #00BB00;">lag()</span>    masks <span style="color: #0000BB;">stats</span>::lag()</span></code></pre></div>
<div class="section level2">
<h2 id="prerequisites">Prerequisites<a class="anchor" aria-label="anchor" href="#prerequisites"></a>
</h2>
<p>In order to pursue simulations in the face of physical linkage, you
must download and install the external dependency, Mendel version 16.
For Windows and Mac OS X, CKMRsim will look for the Mendel binary in its
default install location. So download it from <a href="http://software.genetics.ucla.edu/mendel" class="external-link">http://software.genetics.ucla.edu/mendel</a>
and do a default install. Note that you need to register with your email
in order to download Mendel.</p>
</div>
<div class="section level2">
<h2 id="an-example">An Example<a class="anchor" aria-label="anchor" href="#an-example"></a>
</h2>
<p>‘CKMRsim’ comes with an example data set of linked markers. The data
themselves are concocted—they are merely the microhaplotypes that have
been assigned some fictitious positions within a genome—but that is
useful for demonstration. The data are allele frequencies in the format
required by ‘CKMRsim’. Here are the first few lines of them”</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">linked_mhaps</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 825 × 7</span></span>
<span class="co">#&gt;    Chrom Locus            Pos Allele LocIdx AlleIdx   Freq</span>
<span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>     1 tag_id_2255  2<span style="text-decoration: underline;">778</span>713 AGC         1       1 0.745 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>     1 tag_id_2255  2<span style="text-decoration: underline;">778</span>713 AAC         1       2 0.223 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>     1 tag_id_2255  2<span style="text-decoration: underline;">778</span>713 AGT         1       3 0.032<span style="text-decoration: underline;">6</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>     1 tag_id_1508 10<span style="text-decoration: underline;">014</span>085 GG          2       1 0.956 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>     1 tag_id_1508 10<span style="text-decoration: underline;">014</span>085 AG          2       2 0.022<span style="text-decoration: underline;">0</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>     1 tag_id_1508 10<span style="text-decoration: underline;">014</span>085 GA          2       3 0.022<span style="text-decoration: underline;">0</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>     1 tag_id_2513 11<span style="text-decoration: underline;">890</span>256 CGGAA       3       1 0.621 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>     1 tag_id_2513 11<span style="text-decoration: underline;">890</span>256 CGGAG       3       2 0.121 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>     1 tag_id_2513 11<span style="text-decoration: underline;">890</span>256 CGGGA       3       3 0.089<span style="text-decoration: underline;">5</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">10</span>     1 tag_id_2513 11<span style="text-decoration: underline;">890</span>256 CGGGG       3       4 0.068<span style="text-decoration: underline;">4</span></span>
<span class="co">#&gt; <span style="color: #949494;"># … with 815 more rows</span></span></code></pre></div>
<p>See that each marker is given a position (in base pairs) along the
chromosome upon which it resides.</p>
<p>By default, ‘CKMRsim’ assumes a simple recombination rate of 1 cM per
megabase. If you have an actual genetic linkage map for you markers,
then you can accommodate that by giving base-pair positions
(<code>Pos</code>) of the markers that correspond to the positions that
would yield the observed recombination fractions under a constant rate
of 1 cM/Mb.</p>
<p>Here we will show the process of simulating full sibling pairs with
and without the physical linkage in <code>linked_mhaps</code>. We will
also show that physical linkage does not affect parent-offspring or
unrelated pairs.</p>
</div>
<div class="section level2">
<h2 id="simulating-pairs">Simulating Pairs<a class="anchor" aria-label="anchor" href="#simulating-pairs"></a>
</h2>
<p>Creating the CKMR object is just the same as it was in the “Example
1” vignette.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ck_lmh</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_ckmr.html">create_ckmr</a></span><span class="op">(</span>
  D <span class="op">=</span> <span class="va">linked_mhaps</span>,
  kappa_matrix <span class="op">=</span> <span class="va">kappas</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PO"</span>, <span class="st">"FS"</span>, <span class="st">"U"</span><span class="op">)</span>, <span class="op">]</span>,
  ge_mod_assumed <span class="op">=</span> <span class="va">ge_model_TGIE</span>,
  ge_mod_true <span class="op">=</span> <span class="va">ge_model_TGIE</span>,
  ge_mod_assumed_pars_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>epsilon <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span>,
  ge_mod_true_pars_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>epsilon <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span>
<span class="op">)</span></code></pre></div>
<p>Simulating the markers under the assumption of no linkage is also the
same:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">Qs_lmh_no_link</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_Qij.html">simulate_Qij</a></span><span class="op">(</span><span class="va">ck_lmh</span>,
                               calc_relats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PO"</span>, <span class="st">"FS"</span>, <span class="st">"U"</span><span class="op">)</span>,
                               sim_relats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PO"</span>, <span class="st">"FS"</span>, <span class="st">"U"</span><span class="op">)</span> 
                               <span class="op">)</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: PO</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: FS</span>
<span class="co">#&gt; Simulating unlinked markers from Y_l_true matrices for relationship: U</span></code></pre></div>
<p>In order to simulate the markers with physical linkage, you use the
same <code><a href="../reference/simulate_Qij.html">simulate_Qij()</a></code> function, but you tell it to simulate
with physical linkage <em>and</em> you also have to give it a list of
pedigrees so that it can pass the correct pedigrees to the program
MENDEL, which actually does the simulation. <code>CKMRsim</code> comes
with such a list. At the moment, it only has three entries” “PO”, FS”,
“HS”. If you want to add more (like cousins) feel free. Just note that
the two focal individuals of the pair have to be individuals 1 and
2.<br>
If you need help, ping me at <a href="mailto:eric.anderson@noaa.gov">eric.anderson@noaa.gov</a>.</p>
<p>So, simulating with linkage looks like this:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># we only run this if it is on eric's laptop.  This way</span>
<span class="co"># we can build the vignette even on CRAN's machines that don't</span>
<span class="co"># have Mendel installed</span>
<span class="va">Qs_lmh_with_link</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_Qij.html">simulate_Qij</a></span><span class="op">(</span><span class="va">ck_lmh</span>,
                                 calc_relats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PO"</span>, <span class="st">"FS"</span>, <span class="st">"U"</span><span class="op">)</span>,
                                 sim_relats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PO"</span>, <span class="st">"FS"</span>, <span class="st">"U"</span><span class="op">)</span>, 
                                 unlinked <span class="op">=</span> <span class="cn">FALSE</span>,
                                 pedigree_list <span class="op">=</span> <span class="va">pedigrees</span>
<span class="op">)</span></code></pre></div>
<p>Note that <code>CKMRsim</code> is wise enough to simply simulate
markers as unlinked for the PO, U, and MZ cases (because, when dealing
with pairs with those relationships, the presence of physical linkage
makes no difference to the outcome).</p>
<p>Carrying on, it is instructive to look at the distributions of the
FS/U log-likelihood ratios. They should differ between the cases with
and without linkage, most notably the case with linkage should produce
values with greater variance. Let us see if that is the case. To do
that, we extract the log-likelihood ratio values, put them together in a
single data frame, and then plot them. Note that the
<code>extract_logls</code> function returns a column that says whether
the simulation was done with or without linkage, so it is
straightforward to do this in two lines:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lnl_logls</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>
  <span class="fu"><a href="../reference/extract_logls.html">extract_logls</a></span><span class="op">(</span><span class="va">Qs_lmh_no_link</span>, numer <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>FS <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>, denom <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>U <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span>,
  <span class="fu"><a href="../reference/extract_logls.html">extract_logls</a></span><span class="op">(</span><span class="va">Qs_lmh_with_link</span>, numer <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>FS <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>, denom <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>U <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span>
<span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>Now compare the cases when the truth is full-sibling:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lnl_logls</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">true_relat</span> <span class="op">==</span> <span class="st">"FS"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">logl_ratio</span>, fill <span class="op">=</span> <span class="va">simtype</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>bins <span class="op">=</span> <span class="fl">70</span>, 
                 position <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/position_identity.html" class="external-link">position_identity</a></span><span class="op">(</span><span class="op">)</span>,
                 alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></code></pre></div>
<p><img src="CKMRsim-simulating-linked-markers_files/figure-html/unnamed-chunk-9-1.png" width="672"></p>
<p>See, logl ratios have the same mean top and bottom.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lnl_logls</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">true_relat</span> <span class="op">==</span> <span class="st">"FS"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">simtype</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>mean_logl_rat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">logl_ratio</span><span class="op">)</span>,
            mean_numer <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">numer_logl</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 3</span></span>
<span class="co">#&gt;   simtype  mean_logl_rat mean_numer</span>
<span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span> linked            42.4      -<span style="color: #BB0000;">390.</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">2</span> unlinked          42.3      -<span style="color: #BB0000;">391.</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Eric C. Anderson.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
