<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>create matrix C for probability of observed genotypes from microhaplotype data — ge_model_microhap1 • CKMRsim</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="create matrix C for probability of observed genotypes from microhaplotype data — ge_model_microhap1"><meta property="og:description" content="This is intended for the case where the genotypes in question are composed of alleles
that are actually the multi-SNP haplotypes obtained from next generation sequence data.
In other words, all the SNPs occur on a single read and the phase is known because they
are all together on the read.  It allows for a per-locus or a per-SNP-specific sequencing error rate.  The
haplotypes must be named as strings of A, C, G, or T, (though they could be strings of
any characters---the function isn't going to check that!) and for now  we assume that if the
SNPs are multiallelic then genotyping errors to either of the alternate alleles are
equally likely.  Currently assumes that genotyping errors are equally likely in either
direction at a SNP, too."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CKMRsim</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/CKMRsim-developer-notes.html">CKMRsim-developer-notes</a>
    </li>
    <li>
      <a href="../articles/CKMRsim-example-1.html">CKMRsim-example-1</a>
    </li>
    <li>
      <a href="../articles/CKMRsim-example-2-microsatellites.html">CKMRsim-example-2-microsatellites</a>
    </li>
    <li>
      <a href="../articles/CKMRsim-pairwise-relationships.html">CKMRsim-pairwise-relationships</a>
    </li>
    <li>
      <a href="../articles/CKMRsim-simulating-linked-markers.html">Simulating-linked-markers</a>
    </li>
    <li>
      <a href="../articles/CKMRsim-writing-geno-error-funcs.html">CKMRsim-writing-geno-error-functions</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>create matrix C for probability of observed genotypes from microhaplotype data</h1>
    
    <div class="hidden name"><code>ge_model_microhap1.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This is intended for the case where the genotypes in question are composed of alleles
that are actually the multi-SNP haplotypes obtained from next generation sequence data.
In other words, all the SNPs occur on a single read and the phase is known because they
are all together on the read.  It allows for a per-locus or a per-SNP-specific sequencing error rate.  The
haplotypes must be named as strings of A, C, G, or T, (though they could be strings of
any characters---the function isn't going to check that!) and for now  we assume that if the
SNPs are multiallelic then genotyping errors to either of the alternate alleles are
equally likely.  Currently assumes that genotyping errors are equally likely in either
direction at a SNP, too.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">ge_model_microhap1</span><span class="op">(</span></span>
<span>  <span class="va">L</span>,</span>
<span>  miscall_rate <span class="op">=</span> <span class="fl">0.005</span>,</span>
<span>  dropout_rate <span class="op">=</span> <span class="fl">0.005</span>,</span>
<span>  per_snp_rate <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>L</dt>
<dd><p>an element of the list created by <code><a href="long_markers_to_X_l_list.html">long_markers_to_X_l_list</a></code>. Such
an element basically holds the information at a single locus.  The idea here
is that every ge_mod_* function takes in an object like L, and then can
use any piece of information in it about alleles or genotypes to
configure a genotyping error model.</p></dd>


<dt>miscall_rate</dt>
<dd><p>The rate at which microhaplotype alleles are mis-called.  If
option <code>per_snp_rate</code> is TRUE, then this is the rate at which each SNPs might
get miscalled, such that the pverall miscall rate for microhaplotypes with more SNPs
is higher than for microhaplotypes with few SNPs.</p></dd>


<dt>dropout_rate</dt>
<dd><p>Rate of allelic dropout.</p></dd>


<dt>per_snp_rate</dt>
<dd><p>Logical.  If true, then the overall mis-call rate for the
microhaplotype locus is <code>miscall_rate</code> times the number of SNPs in the microhapotype.
The default is FALSE, and it this option is not really recommended.</p></dd>

</dl></div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># here is some example information about a microhap:</span></span></span>
<span class="r-in"><span><span class="va">example_L_microhap</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $freqs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         GCA         ACA         ACG         ATA         GCG         GTA </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 0.543478261 0.260869565 0.086956522 0.065217391 0.038043478 0.005434783 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $geno_freqs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    GCA / GCA    GCA / ACA    GCA / ACG    GCA / ATA    GCA / GCG    GCA / GTA </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2.953686e-01 2.835539e-01 9.451796e-02 7.088847e-02 4.135161e-02 5.907372e-03 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    ACA / ACA    ACA / ACG    ACA / ATA    ACA / GCG    ACA / GTA    ACG / ACG </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6.805293e-02 4.536862e-02 3.402647e-02 1.984877e-02 2.835539e-03 7.561437e-03 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    ACG / ATA    ACG / GCG    ACG / GTA    ATA / ATA    ATA / GCG    ATA / GTA </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1.134216e-02 6.616257e-03 9.451796e-04 4.253308e-03 4.962193e-03 7.088847e-04 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    GCG / GCG    GCG / GTA    GTA / GTA </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1.447306e-03 4.135161e-04 2.953686e-05 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># now we can feed it in to the function with default parameter values.</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Eric C. Anderson.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

