<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ CKMRsim</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CKMRsim</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.2.999</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/CKMRsim-developer-notes.html">CKMRsim-developer-notes</a>
    </li>
    <li>
      <a href="../articles/CKMRsim-example-1.html">CKMRsim-example-1</a>
    </li>
    <li>
      <a href="../articles/CKMRsim-example-2-microsatellites.html">CKMRsim-example-2-microsatellites</a>
    </li>
    <li>
      <a href="../articles/CKMRsim-pairwise-relationships.html">CKMRsim-pairwise-relationships</a>
    </li>
    <li>
      <a href="../articles/CKMRsim-simulating-linked-markers.html">Simulating-linked-markers</a>
    </li>
    <li>
      <a href="../articles/CKMRsim-writing-geno-error-funcs.html">CKMRsim-writing-geno-error-functions</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="CKMRsim.html">CKMRsim</a></code> </p>
        </td>
        <td><p>My package documentation</p></td>
      </tr><tr><td>
          <p><code><a href="Qij_class.html">Qij_class()</a></code> </p>
        </td>
        <td><p>Constructor function for the class Qij</p></td>
      </tr><tr><td>
          <p><code><a href="brook_trout_genos.html">brook_trout_genos</a></code> </p>
        </td>
        <td><p>Microsatellite data from brook trout</p></td>
      </tr><tr><td>
          <p><code><a href="ckmr2colony.html">ckmr2colony()</a></code> </p>
        </td>
        <td><p>make a rudimentary colony input file</p></td>
      </tr><tr><td>
          <p><code><a href="ckmr_class.html">ckmr_class()</a></code> </p>
        </td>
        <td><p>constructor function for the ckmr class</p></td>
      </tr><tr><td>
          <p><code><a href="ckmr_relats.html">ckmr_relats()</a></code> </p>
        </td>
        <td><p>return a vector of the relationships in a ckmr_class object</p></td>
      </tr><tr><td>
          <p><code><a href="combine_miscalls_and_dropouts.html">combine_miscalls_and_dropouts()</a></code> </p>
        </td>
        <td><p>implements the generalized miscall + dropout model</p></td>
      </tr><tr><td>
          <p><code><a href="comp_ind_pairwise.html">comp_ind_pairwise()</a></code> </p>
        </td>
        <td><p>Compute pairwise relationship measures between all individuals in source and one individual in target</p></td>
      </tr><tr><td>
          <p><code><a href="create_ckmr.html">create_ckmr()</a></code> </p>
        </td>
        <td><p>prepare to simulate pairwise relationships from locus data</p></td>
      </tr><tr><td>
          <p><code><a href="create_integer_genotype_matrix.html">create_integer_genotype_matrix()</a></code> </p>
        </td>
        <td><p>turns long format genotypes into a matrix of integer-code genotypes</p></td>
      </tr><tr><td>
          <p><code><a href="example_L_biallelic.html">example_L_biallelic</a></code> </p>
        </td>
        <td><p>An example biallelic locus</p></td>
      </tr><tr><td>
          <p><code><a href="example_L_microhap.html">example_L_microhap</a></code> </p>
        </td>
        <td><p>An example microhaplotype locus</p></td>
      </tr><tr><td>
          <p><code><a href="example_L_microsat.html">example_L_microsat</a></code> </p>
        </td>
        <td><p>An example microsatellite locus</p></td>
      </tr><tr><td>
          <p><code><a href="extract_logls.html">extract_logls()</a></code> </p>
        </td>
        <td><p>Extract log-likelihood ratios and associated values from a Qij object</p></td>
      </tr><tr><td>
          <p><code><a href="find_close_matching_genotypes.html">find_close_matching_genotypes()</a></code> </p>
        </td>
        <td><p>Return every pair of individuals that mismatch at no more than max_miss loci</p></td>
      </tr><tr><td>
          <p><code><a href="flatten_ckmr.html">flatten_ckmr()</a></code> </p>
        </td>
        <td><p>flatten a from/to pair from a ckmr object (see <code>ckmr_class</code>into a long vector of probabilities</p></td>
      </tr><tr><td>
          <p><code><a href="format.Qij.html">format(<i>&lt;Qij&gt;</i>)</a></code> </p>
        </td>
        <td><p>format method for Qij class (to print)</p></td>
      </tr><tr><td>
          <p><code><a href="format.ckmr.html">format(<i>&lt;ckmr&gt;</i>)</a></code> </p>
        </td>
        <td><p>format method for ckmr class (to print)</p></td>
      </tr><tr><td>
          <p><code><a href="ge_model_TGIE.html">ge_model_TGIE()</a></code> </p>
        </td>
        <td><p>implements a simple true-genotype-independent genotyping error model</p></td>
      </tr><tr><td>
          <p><code><a href="ge_model_microhap1.html">ge_model_microhap1()</a></code> </p>
        </td>
        <td><p>create matrix C for probability of observed genotypes from microhaplotype data</p></td>
      </tr><tr><td>
          <p><code><a href="ge_model_microsat1.html">ge_model_microsat1()</a></code> </p>
        </td>
        <td><p>a simple "length-aware" genotyping error model for microsatellites</p></td>
      </tr><tr><td>
          <p><code><a href="general_allele_based_geno_err_model.html">general_allele_based_geno_err_model()</a></code> </p>
        </td>
        <td><p>implements the generalized Allele-Based genotyping error model</p></td>
      </tr><tr><td>
          <p><code><a href="geometric_chromo_lengths.html">geometric_chromo_lengths()</a></code> </p>
        </td>
        <td><p>simulate a distribution of fake chromosome lengths</p></td>
      </tr><tr><td>
          <p><code><a href="hap_obs_prob.html">hap_obs_prob()</a></code> </p>
        </td>
        <td><p>a simple helper function</p></td>
      </tr><tr><td>
          <p><code><a href="id_prepend.html">id_prepend()</a></code> </p>
        </td>
        <td><p>prepend an ID to a mendel input or output file</p></td>
      </tr><tr><td>
          <p><code><a href="index_ab.html">index_ab()</a></code> </p>
        </td>
        <td><p>a function to compute the genotype index from alleles a and b</p></td>
      </tr><tr><td>
          <p><code><a href="insert_C_l_matrices.html">insert_C_l_matrices()</a></code> </p>
        </td>
        <td><p>insert the matrix C_l into each element of the list of X_l matrices</p></td>
      </tr><tr><td>
          <p><code><a href="insert_Y_l_matrices.html">insert_Y_l_matrices()</a></code> </p>
        </td>
        <td><p>compute the Y_l matrices for each locus in a list</p></td>
      </tr><tr><td>
          <p><code><a href="install_mendel.html">install_mendel()</a></code> </p>
        </td>
        <td><p>Download the mendel binary and install it where CKMRpop expects it</p></td>
      </tr><tr><td>
          <p><code><a href="kappas.html">kappas</a></code> </p>
        </td>
        <td><p>A matrix of Cotterman coefficients for a variety of pairwise relationships</p></td>
      </tr><tr><td>
          <p><code><a href="labrador_salar_ngs_msats.html">labrador_salar_ngs_msats</a></code> </p>
        </td>
        <td><p>Next-generation-sequencing-genotyped microsatellites from Labrador Atlantic salmon</p></td>
      </tr><tr><td>
          <p><code><a href="linked_mhaps.html">linked_mhaps</a></code> </p>
        </td>
        <td><p>Microhaplotype data from kelp rockfish <em>Sebastes atrovirens</em> with pretend linkage positions</p></td>
      </tr><tr><td>
          <p><code><a href="locus_specific_pairwise.html">locus_specific_pairwise()</a></code> </p>
        </td>
        <td><p>Return locus-specific pairwise relationship measures between desired pairs of individuals</p></td>
      </tr><tr><td>
          <p><code><a href="loggy_reweight.html">loggy_reweight()</a></code> </p>
        </td>
        <td><p>reweight logs without underflow</p></td>
      </tr><tr><td>
          <p><code><a href="long_markers.html">long_markers</a></code> </p>
        </td>
        <td><p>A long format way of specifying markers.</p></td>
      </tr><tr><td>
          <p><code><a href="long_markers_to_X_l_list.html">long_markers_to_X_l_list()</a></code> </p>
        </td>
        <td><p>convert a  long data frame of marker allele frequencies to a list of X_l matrices</p></td>
      </tr><tr><td>
          <p><code><a href="make_matrix_X_l.html">make_matrix_X_l()</a></code> </p>
        </td>
        <td><p>compute the matrix X_l given allele frequencies and kappa</p></td>
      </tr><tr><td>
          <p><code><a href="markers40.html">markers40</a></code> </p>
        </td>
        <td><p>40 dinucleotide markers for a quick test data set</p></td>
      </tr><tr><td>
          <p><code><a href="markers_on_map.html">markers_on_map</a></code> </p>
        </td>
        <td><p>An example of how to specify markers mapped onto chromosomes</p></td>
      </tr><tr><td>
          <p><code><a href="mc_sample_simple.html">mc_sample_simple()</a></code> </p>
        </td>
        <td><p>sample Q values to get and analyze a sample of Lambdas with simple (non-mixture) hypotheses</p></td>
      </tr><tr><td>
          <p><code><a href="mendelBin.html">mendelBin()</a></code> </p>
        </td>
        <td><p>Given the operating system, check for mendel program in the typical location</p></td>
      </tr><tr><td>
          <p><code><a href="mendel_control_list.html">mendel_control_list()</a></code> </p>
        </td>
        <td><p>return a list of values for a Mendel definitions file for gene dropping</p></td>
      </tr><tr><td>
          <p><code><a href="microhaps.html">microhaps</a></code> </p>
        </td>
        <td><p>100 microhaplotype markers for a quick test data set</p></td>
      </tr><tr><td>
          <p><code><a href="pairwise_geno_id.html">pairwise_geno_id()</a></code> </p>
        </td>
        <td><p>Return every pair of individuals that mismatch at no more than max_miss loci</p></td>
      </tr><tr><td>
          <p><code><a href="pairwise_kin_logl_ratios.html">pairwise_kin_logl_ratios()</a></code> </p>
        </td>
        <td><p>compute pairwise log likelihood ratios between two sets of individuals</p></td>
      </tr><tr><td>
          <p><code><a href="pedigrees.html">pedigrees</a></code> </p>
        </td>
        <td><p>A list of pedigrees specifying different pairwise relationships for MORGAN's genedrop program.</p></td>
      </tr><tr><td>
          <p><code><a href="print.Qij.html">print(<i>&lt;Qij&gt;</i>)</a></code> </p>
        </td>
        <td><p>print a Qij object nicely</p></td>
      </tr><tr><td>
          <p><code><a href="print.ckmr.html">print(<i>&lt;ckmr&gt;</i>)</a></code> </p>
        </td>
        <td><p>print method for ckmr class</p></td>
      </tr><tr><td>
          <p><code><a href="read_mendel_outped.html">read_mendel_outped()</a></code> </p>
        </td>
        <td><p>pick the genotypes out of the Mendel output pedigree file to use to compute Q</p></td>
      </tr><tr><td>
          <p><code><a href="reindex_markers.html">reindex_markers()</a></code> </p>
        </td>
        <td><p>re-index loci and or alleles  given a data frame of markers</p></td>
      </tr><tr><td>
          <p><code><a href="run_mendel.html">run_mendel()</a></code> </p>
        </td>
        <td><p>Run the mendel binary in directory Dir using control file Control</p></td>
      </tr><tr><td>
          <p><code><a href="samp_from_mat.html">samp_from_mat()</a></code> </p>
        </td>
        <td><p>Sample 1 observation from cell probabilities that are columns of a matrix</p></td>
      </tr><tr><td>
          <p><code><a href="sample_linked_genotype_pairs.html">sample_linked_genotype_pairs()</a></code> </p>
        </td>
        <td><p>simulate genotype-pairs from linked markers using Mendel</p></td>
      </tr><tr><td>
          <p><code><a href="sebastes.html">sebastes</a></code> </p>
        </td>
        <td><p>Microhaplotype data from kelp rockfish <em>Sebastes atrovirens</em></p></td>
      </tr><tr><td>
          <p><code><a href="simulate_Qij.html">simulate_Qij()</a></code> </p>
        </td>
        <td><p>compute the all-locus genotype pair probs for pairs simulated given specified relationships</p></td>
      </tr><tr><td>
          <p><code><a href="simulate_and_calc_Q.html">simulate_and_calc_Q()</a></code> </p>
        </td>
        <td><p>compute all-locus genotype pair probs for pairs simulated given specified relationships</p></td>
      </tr><tr><td>
          <p><code><a href="sprinkle_markers_into_genome.html">sprinkle_markers_into_genome()</a></code> </p>
        </td>
        <td><p>Randomly distribute markers along the chromosomes of a genome</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Eric C. Anderson.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

