<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>convert a  long data frame of marker allele frequencies to a list of X_l matrices — long_markers_to_X_l_list • CKMRsim</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="convert a  long data frame of marker allele frequencies to a list of X_l matrices — long_markers_to_X_l_list" />
<meta property="og:description" content="Given a data frame D in the format of long_markers, this function creates the matrix X_l for
each locus, given all the values of kappa supplied as rows in a matrix like that in
kappas. The rows and columns refer to genotypes that are in
the order given in the
function index_ab." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CKMRsim</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/CKMRsim-developer-notes.html">CKMRsim-developer-notes</a>
    </li>
    <li>
      <a href="../articles/CKMRsim-example-1.html">CKMRsim-example-1</a>
    </li>
    <li>
      <a href="../articles/CKMRsim-example-2-microsatellites.html">CKMRsim-example-2-microsatellites</a>
    </li>
    <li>
      <a href="../articles/CKMRsim-pairwise-relationships.html">CKMRsim-pairwise-relationships</a>
    </li>
    <li>
      <a href="../articles/CKMRsim-simulating-linked-markers.html">Simulating-linked-markers</a>
    </li>
    <li>
      <a href="../articles/CKMRsim-writing-geno-error-funcs.html">CKMRsim-writing-geno-error-functions</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>convert a  long data frame of marker allele frequencies to a list of X_l matrices</h1>
    
    <div class="hidden name"><code>long_markers_to_X_l_list.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Given a data frame D in the format of long_markers, this function creates the matrix X_l for
each locus, given all the values of kappa supplied as rows in a matrix like that in
<code><a href='kappas.html'>kappas</a></code>. The rows and columns refer to genotypes that are in
the order given in the
function <code><a href='index_ab.html'>index_ab</a></code>.</p>
    </div>

    <pre class="usage"><span class='fu'>long_markers_to_X_l_list</span><span class='op'>(</span><span class='va'>D</span>, <span class='va'>kappa_matrix</span>, allele_separator <span class='op'>=</span> <span class='st'>" / "</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>D</th>
      <td><p>a data frame in the format of <code><a href='long_markers.html'>long_markers</a></code>.  Monomorphic loci should have
been removed and the whole thing run through <code><a href='reindex_markers.html'>reindex_markers</a></code> before passing it
to this function.</p></td>
    </tr>
    <tr>
      <th>kappa_matrix</th>
      <td><p>A matrix like that in the supplied data <code><a href='kappas.html'>kappas</a></code>.  It should have
three columns, and rownames giving the abbreviated name of the pairwise relationship.  Each row
is a three-vector which are the Cotterman coefficients for the relationship. The first element is the
probability that the pair shares 0 gene copies IBD, the second is the prob that they share 1 gene
copy IBD, and the third is the prob that they share 2 gene copies IBD, all assuming no inbreeding.</p></td>
    </tr>
    <tr>
      <th>allele_separator</th>
      <td><p>This string is used to separate allele names in the names of genotypes.
By default it is " / ".  This should not get confused by any elements in actual allele names, as
long as the allele names don't have any spaces in them.  This gets used to parse the genotype
names down the road.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>This returns a named list.  The names of the components are the Chrom.Locus.Pos of the
marker in D. The contents of each list component is a list that includes the allele frequencies
(as a vector named with the Allele names), and also another list of matrices with nA * (nA+1) / 2 rows and
columns.  The rows and columns of this matrix are named by the genotypes.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>kappas</span><span class='op'>)</span>
<span class='va'>lm_example</span> <span class='op'>&lt;-</span> <span class='fu'>long_markers_to_X_l_list</span><span class='op'>(</span><span class='va'>long_markers</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>18</span>,<span class='op'>]</span>, kappa_matrix <span class='op'>=</span> <span class='va'>kappas</span><span class='op'>)</span>
<span class='va'>mh_example</span> <span class='op'>&lt;-</span> <span class='fu'>long_markers_to_X_l_list</span><span class='op'>(</span><span class='va'>microhaps</span>, kappa_matrix <span class='op'>=</span> <span class='va'>kappas</span><span class='op'>)</span>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Eric C. Anderson.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


