<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>create matrix C for probability of observed genotypes from microhaplotype data — microhaplotype_geno_err_matrix • CKMRsim</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="create matrix C for probability of observed genotypes from microhaplotype data — microhaplotype_geno_err_matrix"><meta property="og:description" content="This is intended for the case where the genotypes in question are composed of alleles
that are actually the multi-SNP haplotypes obtained from next generation sequence data.
In other words, all the SNPs occur on a single read and the phase is known because they
are all together on the read.  It allows for a SNP-specific sequencing error rate.  The
haplotypes must be named as strings of A, C, G, or T, (though they could be strings of
any characters---the function isn't going to check that!) and for now  we assume that if the
SNPs are multiallelic then genotyping errors to either of the alternate alleles are
equally likely.  Currently assumes that genotyping errors are equally likely in either
direction at a SNP, too."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CKMRsim</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/CKMRsim-developer-notes.html">CKMRsim-developer-notes</a>
    </li>
    <li>
      <a href="../articles/CKMRsim-example-1.html">CKMRsim-example-1</a>
    </li>
    <li>
      <a href="../articles/CKMRsim-example-2-microsatellites.html">CKMRsim-example-2-microsatellites</a>
    </li>
    <li>
      <a href="../articles/CKMRsim-pairwise-relationships.html">CKMRsim-pairwise-relationships</a>
    </li>
    <li>
      <a href="../articles/CKMRsim-simulating-linked-markers.html">Simulating-linked-markers</a>
    </li>
    <li>
      <a href="../articles/CKMRsim-writing-geno-error-funcs.html">CKMRsim-writing-geno-error-functions</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>create matrix C for probability of observed genotypes from microhaplotype data</h1>
    
    <div class="hidden name"><code>microhaplotype_geno_err_matrix.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This is intended for the case where the genotypes in question are composed of alleles
that are actually the multi-SNP haplotypes obtained from next generation sequence data.
In other words, all the SNPs occur on a single read and the phase is known because they
are all together on the read.  It allows for a SNP-specific sequencing error rate.  The
haplotypes must be named as strings of A, C, G, or T, (though they could be strings of
any characters---the function isn't going to check that!) and for now  we assume that if the
SNPs are multiallelic then genotyping errors to either of the alternate alleles are
equally likely.  Currently assumes that genotyping errors are equally likely in either
direction at a SNP, too.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">microhaplotype_geno_err_matrix</span><span class="op">(</span></span>
<span>  <span class="va">haps</span>,</span>
<span>  snp_err_rates <span class="op">=</span> <span class="fl">0.005</span>,</span>
<span>  dropout_rates <span class="op">=</span> <span class="fl">0.005</span>,</span>
<span>  scale_by_num_snps <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>haps</dt>
<dd><p>character vector of strings that denote the haplotypes at the locus.  For example
"CCAG", "CTAG", "GCAG", etc.  Note that these should be in the same order as they are given
in the allele frequency definitions (so that the ordering of genotypes made from them will
be correct).  Each element of haps must be a string of the same number
of characters.  haps cannot be a factor.</p></dd>


<dt>snp_err_rates</dt>
<dd><p>Vector of rates at which sequencing errors are expected at each of the SNPs
that are in the haplotype.  This recycles if its length is less than the number of SNPs in the
haplotypes.</p></dd>


<dt>dropout_rates</dt>
<dd><p>Haplotype-specific rates of allelic dropout.  Recycles if need be.</p></dd>


<dt>scale_by_num_snps</dt>
<dd><p>Logical.  If true, then the error rate is divided by the
number of SNPs in each microhaplotype.</p></dd>

</dl></div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># five haplotypes in alphabetical order</span></span></span>
<span class="r-in"><span><span class="va">haps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"AACC"</span>, <span class="st">"GACC"</span>, <span class="st">"GATA"</span>, <span class="st">"GTCC"</span>, <span class="st">"GTTC"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># make the matrix C</span></span></span>
<span class="r-in"><span><span class="va">C_mat</span> <span class="op">&lt;-</span> <span class="fu">microhaplotype_geno_err_matrix</span><span class="op">(</span><span class="va">haps</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># look at the first part of it</span></span></span>
<span class="r-in"><span><span class="va">C_mat</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>            obs_geno</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> true_geno     AACC-AACC    AACC-GACC    AACC-GATA    AACC-GTCC    AACC-GTTC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   AACC-AACC 0.990024753 9.850250e-03 2.462563e-07 4.925125e-05 2.462563e-07</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   AACC-GACC 0.009924999 9.751503e-01 2.462624e-05 4.925249e-03 2.462624e-05</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   AACC-GATA 0.004974998 2.462624e-05 9.849755e-01 1.231312e-07 2.462563e-05</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   AACC-GTCC 0.004999624 4.925249e-03 1.231312e-07 9.751500e-01 4.925125e-03</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   AACC-GTTC 0.004974998 2.462624e-05 2.462563e-05 4.925125e-03 9.800505e-01</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Eric C. Anderson.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

